[hip] Docs: Remove nvcc references (#2922)
ROCm/rocm-systems: projects/hip/docs/how-to/hip_porting_guide.rst

.. meta::
  :description: This chapter presents how to port NVIDIA CUDA source code to HIP
  :keywords: AMD, ROCm, HIP, NVIDIA CUDA, driver API, porting, port

.. _porting_cuda_code:

*******************************************************************************
Porting NVIDIA CUDA code to HIP
*******************************************************************************

HIP is a C++ runtime API and kernel language for AMD GPUs. It allows developers to convert
existing NVIDIA CUDA code to run on AMD GPUs. This topic describes the available tools and

> provides practical suggestions for porting your NVIDIA CUDA code and working through common issues.

Reviewer: ```suggestion
HIP is a C++ runtime API and kernel language for AMD GPUs. It allows developers to convert
existing NVIDIA CUDA code to run on AMD GPUs. This topic describes the available tools and
provides practical suggestions for porting your CUDA code and working through common issues.
```

Reviewer: The first occurrence on the page needs NVIDIA. The rest can go without.

NVIDIA CUDA provides separate driver and runtime APIs, while HIP mostly uses a single API.
The two NVIDIA CUDA APIs generally provide similar functionality and are mostly interchangeable.

> However, the NVIDIA CUDA driver API provides fine-grained control over kernel-level

Reviewer: ```suggestion
CUDA provides separate driver and runtime APIs, while HIP mostly uses a single API.
The two CUDA APIs generally provide similar functionality and are mostly interchangeable.
However, the CUDA driver API provides fine-grained control over kernel-level
```

prefixed with ``hipDrv``.

Porting a CUDA project
=============================

HIP provides a runtime API for AMD GPUs that closely mirrors CUDA, making it
straightforward to convert CUDA applications. To compile HIP code for AMD platforms,
you can use ``amdclang++``, also called HIP-Clang, or you can use ``hipcc`` which provides

> a compiler interface.

Reviewer 2: ```suggestion
HIP provides a runtime API for AMD GPUs that closely mirrors CUDA, making it straightforward to port existing CUDA applications. To compile HIP code for AMD platforms, you can use ``amdclang++`` (also known as HIP-Clang) or ``hipcc``, which offers a higher‑level compiler interface.
```

General tips
------------

* Use the `HIPIFY <https://github.com/ROCm/HIPIFY>`_ tools to automatically
  convert CUDA code to HIP, as described in the following section.
* Start with a working CUDA codebase before beginning the porting process.

> * Port incrementally, testing functionality as you convert code sections.

Reviewer 2: ```suggestion
* Port the application incrementally and test each section as you convert it.
```

> * HIP code runs on AMD GPUs and takes advantage of AMD's ROCm software stack.

Reviewer 2: ```suggestion
* HIP code runs on AMD GPUs and takes advantage of the ROCm software stack for both performance and tooling support.
```

* :doc:`hipify-clang <hipify:how-to/hipify-clang>` is a Clang-based tool that parses code,
  translates it into an Abstract Syntax Tree, and generates the HIP source. For this,
  ``hipify-clang`` needs to be able to actually compile the code, so the NVIDIA CUDA code needs
  to be correct, and a NVIDIA CUDA install with all necessary headers must be provided.

* :doc:`hipify-perl <hipify:how-to/hipify-perl>` uses pattern matching, to translate the
  NVIDIA CUDA code to HIP. It does not require a working NVIDIA CUDA installation, and can also
  convert NVIDIA CUDA code, that is not syntactically correct. It is therefore easier to

>   set up and use, but is not as powerful as ``hipfiy-clang``.

Reviewer: ```suggestion
  set up and use, but is not as powerful as ``hipifiy-clang``.
```

HIP-Clang defines a process-wide address space where
the CPU and all devices allocate addresses from a single unified pool.
This means addresses can be shared between contexts. A new context
does not create a new address space for the device.

Context stack behavior differences
----------------------------------

HIP-Clang creates a primary context when the HIP API is called.
HIP-Clang pushes the primary context to the context stack when it
is empty. This behavior differs from the CUDA driver API code where

> contexts may need to be explicitly managed.

Reviewer 2: ```suggestion
HIP-Clang creates a primary context when the HIP API is first invoked. It then pushes this primary context onto the context stack if the stack is empty. This behavior differs from the CUDA driver API, where contexts often need to be managed explicitly.
```

HIP-Clang defines a process-wide address space where
the CPU and all devices allocate addresses from a single unified pool.
This means addresses can be shared between contexts. A new context
does not create a new address space for the device.

Context stack behavior differences
----------------------------------

HIP-Clang creates a primary context when the HIP API is called.
HIP-Clang pushes the primary context to the context stack when it
is empty. This behavior differs from the NVIDIA CUDA driver API code where
contexts may need to be explicitly managed.

Scanning NVIDIA CUDA source to scope the translation

> ---------------------------------------------

Reviewer 3: ```suggestion
Scanning NVIDIA CUDA source to scope the translation
----------------------------------------------------
```

    cudaMemcpyToSymbol => hipMemcpyToSymbol: 1
    cuda_runtime.h => hip/hip_runtime.h: 1

``hipexamine-perl.sh`` reports how many NVIDIA CUDA calls are going to be converted to
HIP (e.g. ``CONVERTED refs count: 2``), and lists them by name together with
their corresponding HIP-version (see the lines following ``[HIPIFY] info:
CONVERTED refs by names:``). It also lists the total lines of code for the file
and potential warnings. In the end it prints a summary for all files.

Automatically converting a NVIDIA CUDA project

> ---------------------------------------

Reviewer 3: ```suggestion
Automatically converting a NVIDIA CUDA project
----------------------------------------------
```

Most CUDA libraries have a corresponding HIP library. For more information,
see either :doc:`ROCm libraries <rocm:reference/api-libraries>` or :doc:`HIPIFY CUDA compatible libraries <hipify:reference/supported_apis>`.

There are two flavors of libraries provided by ROCm, libraries prefixed with ``hip``
and libraries prefixed with ``roc``. While both are written using HIP, libraries
with the ``roc``-prefix are optimized for AMD GPUs and may use AMD-specific

> features to achieve the best performance.

Reviewer 2: ```suggestion
ROCm provides two categories of libraries: those prefixed with ``hip`` and those prefixed with ``roc``. While both are implemented using HIP, the ``roc`` libraries are optimized specifically for AMD GPUs and may use AMD‑specific features to deliver the best performance.
```

current context or to query properties of the device associated with the context. The current
context is implicitly used by other APIs, such as ``hipStreamCreate``.

.. note::
  The ``hipCtx`` API is **deprecated** and its use is discouraged. Most new applications use
  ``hipSetDevice`` or the ``hipStream`` APIs. For more details on deprecated APIs, see :doc:`../reference/deprecated_api_list`.

.. _compilation_platform:

Compilation
===========

HIP code must be compiled to target AMD GPU architectures. The created binaries

> contain code specific to the target GPU architecture.

Reviewer 2: ```suggestion
HIP code must be compiled for a specific AMD GPU architecture, and the resulting binaries contain code tailored to that target architecture.
```

``hipcc`` is a compiler driver that calls ``amdclang++`` (HIP-Clang),
passing the necessary options to the compiler. Tools that call ``hipcc``

> must ensure the compiler options are appropriate for the target compiler.

Reviewer 2: ```suggestion
``hipcc`` is a compiler driver that invokes ``amdclang++`` (HIP-Clang) and passes the required options to it. Tools that rely on ``hipcc`` must ensure that the compiler flags they provide are appropriate for the underlying compiler.
```

  hipconfig --cpp_config
   -D__HIP_PLATFORM_AMD__= -I/opt/rocm/include

HIP-Clang does not include default headers, and instead you must explicitly
include all required files.

.. note::
  The ``hipify`` tool automatically converts ``cuda_runtime.h`` to ``hip_runtime.h``,
  and it converts ``cuda_runtime_api.h`` to ``hip_runtime_api.h``, but it may
  miss nested headers or macros.

Compiler defines for HIP
-------------------------

> C++-macros are defined by the HIP compilers and APIs. This section lists macros

Reviewer 3: ```suggestion
C++ macros are defined by the HIP compilers and APIs. This section lists macros
```

warpSize
========

Code should not assume a warp size of 32 or 64, as AMD GPU architectures have different
warp sizes. The ``warpSize`` built-in should be used in device code, while the host can
query it during runtime via the device properties. See the :ref:`HIP language extension

> for warpSize <warp_size>` for information on how to write warpSize-aware code.

Reviewer 3: ```suggestion
query it during runtime via the device properties. See the :ref:`HIP language extension for warpSize warp_size>` for information on how to write warpSize-aware code.
```

I don't think the reference will be correctly parsed across a line break.
